File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SchemaCodeGenUtils.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.ui.wizard

// TODO: This is fragile. Very fragile. But it is necessary to get Schemas service launched, and we've evaluated all other trade offs
// This will be done on the server-side as soon as we can, but for now the client needs to do this
class SchemaCodeGenUtils {
    companion object {
        private const val SCHEMA_PACKAGE_PREFIX = "schema"
        private const val AWS = "aws"
        private const val PARTNER = "partner"
        private const val AWS_PARTNER_PREFIX = "$AWS.$PARTNER-" // dash suffix because of 3p partner registry name format
        private const val AWS_EVENTS_PREFIX = "$AWS." // . suffix because of 1p event registry schema format

        fun buildSchemaPackageName(schemaName: String): String {
            val builder = CodeGenPackageBuilder()
            builder.append(SCHEMA_PACKAGE_PREFIX)
            buildPackageName(builder, schemaName)
            return builder.toString()
        }

        private fun buildPackageName(builder: CodeGenPackageBuilder, schemaName: String) {
            if (isAwsPartnerEvent(schemaName)) {
                buildPartnerEventPackageName(builder, schemaName)
            } else if (isAwsEvent(schemaName)) {
                buildAwsEventPackageName(builder, schemaName)
            } else {
                buildCustomPackageName(builder, schemaName)
            }
        }

        private fun isAwsPartnerEvent(schemaName: String): Boolean = schemaName.startsWith(AWS_PARTNER_PREFIX)

        private fun buildPartnerEventPackageName(builder: CodeGenPackageBuilder, schemaName: String) {
            val partnerSchemaString = schemaName.substring(AWS_PARTNER_PREFIX.length)

            builder
                .append(AWS)
                .append(PARTNER)
                .append(partnerSchemaString)
        }

        private fun isAwsEvent(name: String): Boolean = name.startsWith(AWS_EVENTS_PREFIX)

        private fun buildAwsEventPackageName(builder: CodeGenPackageBuilder, schemaName: String) {
            val awsEventSchemaParts = schemaName.split(".")
            for (part in awsEventSchemaParts) {
                builder.append(part)
            }
        }

        private fun buildCustomPackageName(builder: CodeGenPackageBuilder, schemaName: String) = builder.append(schemaName)
    }

    class CodeGenPackageBuilder {
        private val builder: StringBuilder = StringBuilder()

        fun append(segment: String): CodeGenPackageBuilder {
            if (builder.isNotEmpty()) {
                builder.append(IdentifierFormatter.PACKAGE_SEPARATOR)
            }
            builder.append(IdentifierFormatter.toValidIdentifier(segment.toLowerCase()))
            return this
        }

        override fun toString(): String = builder.toString()
    }

    class IdentifierFormatter {
        companion object {
            private const val POTENTIAL_PACKAGE_SEPARATOR = "@"

            private const val NOT_VALID_IDENTIFIER_CHARACTER = "[^a-zA-Z0-9_$POTENTIAL_PACKAGE_SEPARATOR]"
            private val NOT_VALID_IDENTIFIER_REGEX = Regex(NOT_VALID_IDENTIFIER_CHARACTER)

            const val PACKAGE_SEPARATOR = "."

            private const val UNDERSCORE = "_"

            fun toValidIdentifier(name: String): String = name
                .replace(NOT_VALID_IDENTIFIER_REGEX, UNDERSCORE)
                .replace(POTENTIAL_PACKAGE_SEPARATOR, PACKAGE_SEPARATOR)
        }
    }
}

File: src/eventSchemas/models/schemaCodeGenUtils.ts
/*!
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

// TODO: This is fragile. Very fragile. But it is necessary to get Schemas service launched, and we've evaluated all other tradeoffs
// This will be done on the server-side as soon as we can, but for now the client needs to do this
export class SchemaCodeGenUtils {
    private readonly SCHEMA_PACKAGE_PREFIX = 'schema'
    private readonly AWS = 'aws'
    private readonly PARTNER = 'partner'
    private readonly AWS_PARTNER_PREFIX = `${this.AWS}.${this.PARTNER}-` // dash suffix because of 3p partner registry name format
    private readonly AWS_EVENTS_PREFIX = `${this.AWS}.` // . suffix because of 1p event registry schema format

    public buildSchemaPackageName(schemaName: string): string {
        const builder = new CodeGenPackageBuilder()
        builder.append(this.SCHEMA_PACKAGE_PREFIX)
        this.buildPackageName(builder, schemaName)

        return builder.build()
    }

    private buildPackageName(builder: CodeGenPackageBuilder, schemaName: string): void {
        // do not modify the order of conditional checks
        if (this.isAwsPartnerEvent(schemaName)) {
            this.buildPartnerEventPackageName(builder, schemaName)
        } else if (this.isAwsEvent(schemaName)) {
            this.buildAwsEventPackageName(builder, schemaName)
        } else {
            this.buildCustomPackageName(builder, schemaName)
        }
    }

    private isAwsPartnerEvent(schemaName: string): boolean {
        return schemaName.startsWith(this.AWS_PARTNER_PREFIX)
    }

    private buildPartnerEventPackageName(builder: CodeGenPackageBuilder, schemaName: string): void {
        const partnerSchemaString = schemaName.substring(this.AWS_PARTNER_PREFIX.length)

        builder.append(this.AWS).append(this.PARTNER).append(partnerSchemaString)
    }

    private isAwsEvent(name: string): boolean {
        return name.startsWith(this.AWS_EVENTS_PREFIX)
    }

    private buildAwsEventPackageName(builder: CodeGenPackageBuilder, schemaName: string): void {
        const awsEventSchemaParts = schemaName.split('.')
        for (const part of awsEventSchemaParts) {
            builder.append(part)
        }
    }

    private buildCustomPackageName(builder: CodeGenPackageBuilder, schemaName: string): void {
        builder.append(schemaName)
    }
}

class CodeGenPackageBuilder {
    private builder = ''
    public build(): string {
        return this.builder
    }

    public append(segment: string): CodeGenPackageBuilder {
        if (this.builder.length > 0) {
            this.builder = this.builder.concat(IdentifierFormatter.PACKAGE_SEPARATOR)
        }
        this.builder = this.builder.concat(IdentifierFormatter.toValidIdentifier(segment.toLowerCase()))

        return this
    }
}

export namespace IdentifierFormatter {
    export const PACKAGE_SEPARATOR = '.'
    const POTENTIAL_PACKAGE_SEPARATOR = '@'
    const NOT_VALID_IDENTIFIER_REGEX = new RegExp(`[^a-zA-Z0-9_${POTENTIAL_PACKAGE_SEPARATOR}]`, 'g')
    const POTENTIAL_PACKAGE_SEPARATOR_REGEX = new RegExp(POTENTIAL_PACKAGE_SEPARATOR, 'g')
    const UNDERSCORE = '_'

    export function toValidIdentifier(name: string): string {
        return name
            .replace(NOT_VALID_IDENTIFIER_REGEX, UNDERSCORE)
            .replace(POTENTIAL_PACKAGE_SEPARATOR_REGEX, PACKAGE_SEPARATOR)
    }
}
